### DISCUSSION TITLE
Max heap size keeps filling up

### DISCUSSION THREAD

**Discussion by marcoamujica** [DependencyTrack/dependency-track#2222] [Q&A] [answered]
Hi, everyone.  I deployed Dtrack using the docker-compose.yml referenced in the documentation.  After a month or so, the Java max heap size filled up.  I increased it from the default 3G to 5G, but it managed to fill up again in just a couple of days.  I was wondering what would be the best way to deal with this.  Are we just pushing the tool past what its intended use is?  Is this due to a misconfiguration or a bug?  Thanks in advance, everybody.

**Comment by nscuro** [comment, marked as answer]
Are you experiencing any performance or stability issues, or is it just the heap space increasing that makes you uncomfortable?

The JVM will take whatever space you make available to it, that is expected behavior to some extent.

**Reply by marcoamujica** [reply]
The heap space fills up and crashes the service, causing 404 errors.  I recently increased the max heap size to 12G, but that seems like an unreasonable increase from the default 3G.

**Reply by nscuro** [reply]
If I may ask, what is your current workload? As in, how many projects and components do you track with DT? How frequent are you uploading SBOMs? 

Is it possible for you to correlate increases in memory usage to certain events, e.g. bursts of SBOM uploads, some recurring tasks running etc.?

**Reply by marcoamujica** [reply]
It looks like we have 1,229 rows in the projects tab.  I'll find out about the SBOM upload rate.

**Reply by nscuro** [reply]
If you have direct DB access, this should yield the number of components in active projects:

```sql
SELECT COUNT(*) FROM "COMPONENT" AS "C" WHERE (SELECT COUNT(*) FROM "PROJECT" AS "P" WHERE "P"."ID" = "C"."PROJECT_ID" AND "P"."ACTIVE" = true) > 0
```

Would be helpful to know, as only components in active projects are being processed by the various recurring tasks.

**Reply by marcoamujica** [reply]
Ah, right.  Thank you for all the quick responses.  I'll check now.

**Reply by marcoamujica** [reply]
I'm not getting any results from the command.
`dtrack=# SELECT COUNT(*) FROM "COMPONENT" AS "C" WHERE (SELECT COUNT(*) FROM "PROJECT" AS "P" WHERE "P"."ID" = "C"."PROJECT_ID" AND "P"."ACTIVE" = true) > 0`
`dtrack-#`
When I look in the web ui, sorting by the Active column only shows active projects.  Selecting and de-selecting the Show Inactive Projects slider yields the same amount of projects, 1,234 now.

**Reply by marcoamujica** [reply]
After talking with developers, it sounds like we're uploading way too many SBOMs.  "SBOMs are currently generated for all branches (feature/task/PR/main(develop) ) and all pipelines (Java Dotnet, docker images)."  Limiting to PR and main should help out a lot.  Currently looking for the best way to monitor resource usage to track how effective this will be.

**Reply by nscuro** [reply]
DT exposes Prometheus metrics, which includes metrics for the host and the JVM: https://docs.dependencytrack.org/getting-started/monitoring/

**Reply by marcoamujica** [reply]
Ah, perfect.  Thank you.

**Reply by marcoamujica** [reply]
I figure I'll mark this solved based on the likelihood that we're just generating way too much work for the API server.  I'll track metrics and see how things pan out.  Thank you so much for all your help, @nscuro!

**Comment by marcoamujica** [comment]
I was hoping that pairing down SBOM submission to just PR and prod branches would do the trick, but heap is still filling up at an alarming rate, causing the API container to crash.  We're up to 1,474 active projects.  Are all of these active projects constantly stealing resources?  If so, I'm wondering what would be the best way to bulk delete projects and how to keep from creating so many projects in the future.
